impersonate_login/models/res_users.py

```diff
@@ -0,0 +1,118 @@
+                            "impersonated_partner_id": self.env["res.users"]
+                            .browse(target_uid)
+                            .partner_id.id,
+                            "date_start": fields.datetime.now(),
+                        }
+                    )
+                )
+                request.session.impersonate_log_id = impersonate_log.id
+                logger.info(
+                    f"IMPERSONATE: {self._get_partner_name(self._uid)} "
+                    f"Login as {self._get_partner_name(self.id)}"
+                )
+                # invalidate session token cache as we've changed the uid
+                request.env.registry.clear_cache()
+                request.session.session_token = security.compute_session_token(
+                    request.session, request.env
+                )
+
+                # reload the client; open the first available root menu
+                menu = self.env["ir.ui.menu"].search([("parent_id", "=", False)])[:1]
+                return {
+                    "type": "ir.actions.client",
+                    "tag": "reload",
+                    "params": {"menu_id": menu.id},
+                }
+
+    @api.model
+    def action_impersonate_login(self):
+        if request:
+            from_uid = request.session.impersonate_from_uid
+            if not from_uid:
+                action = self.env["ir.actions.act_window"]._for_xml_id(
+                    "base.action_res_users"
+                )
+                action["views"] = [[self.env.ref("base.view_users_tree").id, "tree"]]
+                action["domain"] = [
+                    ("id", "!=", self.env.user.id),
+                    ("share", "=", False),
+                ]
+                action["target"] = "new"
+                return action
+
+    @api.model
+    def back_to_origin_login(self):
+        if request:
+            from_uid = request.session.impersonate_from_uid
+            if from_uid:
+                request.session.uid = from_uid
+                self.env["impersonate.log"].sudo().browse(
+                    request.session.impersonate_log_id
+                ).write(
+                    {
+                        "date_end": fields.datetime.now(),
+                    }
+                )
+                # invalidate session token cache as we've changed the uid
+                request.env.registry.clear_cache()
+                request.session.impersonate_from_uid = False
+                request.session.impersonate_log_id = False
+                request.session.session_token = security.compute_session_token(
+                    request.session, request.env
+                )
+                logger.info(
+                    f"IMPERSONATE: {self._get_partner_name(from_uid)} "
+                    f"Logout as {self._get_partner_name(self._uid)}"
+                )
+            return {
+                "type": "ir.actions.client",
+                "tag": "reload",
```

Comment:
It is possible to get an access error when going back to the original user as well! Would you consider redirecting to the main menu here as well? 

![image](https://github.com/user-attachments/assets/a2235cd0-77e0-4b0e-828f-b4988ec74b69)